### PR TITLE
3.4 port spatial index to use same pattern as temporal index

### PR DESCRIPTION
**Pull request by OliviaYtterbrink**
The `loadAll` method should be worked on.

~~This PR builds on https://github.com/neo4j/neo4j/pull/11181~~

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCache.java [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+ * Iterating over the cache will return all currently created parts.
+ *
+ * @param <T> Type of parts
+ * @param <E> Type of exception potentially thrown during creation
+ */
+class SpatialIndexCache<T, E extends Exception> implements Iterable<T>
+{
+    private final Factory<T, E> factory;
+
+    private ConcurrentMap<CoordinateReferenceSystem,T> spatials = new ConcurrentHashMap<>();
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCache.java]: I think a copy-on-write map would be appropriate here, as we will write very rarely.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexPopulatingUpdater.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        PointValue value = (PointValue) update.values()[0];
+        switch ( update.updateMode() )
+        {
+        case ADDED:
+            select( value.getCoordinateReferenceSystem() ).process( update );
+            break;
+
+        case CHANGED:
+            // These are both temporal, but could belong in different parts
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexPopulatingUpdater.java]: temporal?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexReader.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        query( nodeValueIterator, IndexOrder.NONE, predicates );
+        return nodeValueIterator;
+    }
+
+    @Override
+    public void query( IndexProgressor.NodeValueClient cursor, IndexOrder indexOrder, IndexQuery... predicates )
+    {
+        if ( predicates.length != 1 )
+        {
+            throw new IllegalArgumentException( "Only single property temporal indexes are supported." );
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexReader.java]: temporal again

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexReader.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+    }
+
+    @Override
+    public IndexSampler createSampler()
+    {
+        return new FusionIndexSampler( Iterators.stream( iterator() ).map( IndexReader::createSampler ).toArray( IndexSampler[]::new ) );
+    }
+
+    @Override
+    public PrimitiveLongResourceIterator query( IndexQuery... predicates )
```
Comment (Lojjs) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexReader.java:73]: Both the old implementation and the temporal version have a special case for `predicates[0] instanceof ExisitsPredicate` here. Was it obsolete due to the special handling of the same case in the `query` method below or why was it removed?

**Review by Lojjs — APPROVED** [review]
Great work refactoring this. I really like that we follow the same pattern as for temporal now, which I find less confusing. Maybe there is a bit too much duplicate code between the spatial and temporal solutions now. One idea for future improvement would be to let them have a common superclass e.g. SpatialIndexAccessor extends XxxIndexAccessor implements IndexAccessor. You also have removed a lot of test files, I guess the temporal test files are considered to be covering both solutions? If we choose to create a common superclass later we could of course rename them so its clearer that they test all implementations/extensions of the superclass and not only the temporal one.

**Pull request merged**